### PR TITLE
Re-enable the Coconizer loader

### PR DESCRIPTION
**Pull request by ccawley2011**
The Fine Volume Up, Fine Volume Down and Set Stereo Position effects have been fixed, but a number of commands are still unimplemented (0x08, 0x09, 0x0a, 0x0b, 0x11, 0x12) or handled incorrectly (0x05, 0x06), and I'm not sure how to implement them properly. Sample files are available from [APDL](http://www.apdl.org.uk/apdlpd/library/sounds.htm).

### CODE REVIEW

**Comment by debrouxl** [comment]
I'm not a maintainer, but have you fuzzed this implementation, using a standard build of libxmp+xmp instrumented by AddressSanitizer + AFL++/Honggfuzz, or a special build of libxmp instrumented by AddressSanitizer + the custom minimal front-end built by libfuzzer sketched in #302 ? :)

**Comment by sezero** [comment]
>  but have you fuzzed this implementation

The loader at least needs some sanity checks.  A _possible_ patch (which
might not be correct or might very possibly be incomplete) on top of this
changeset is below.
```diff
diff --git a/src/loaders/coco_load.c b/src/loaders/coco_load.c
index 16c58aa..1f26ced 100644
--- a/src/loaders/coco_load.c
+++ b/src/loaders/coco_load.c
@@ -58,7 +58,7 @@ static int coco_test(HIO_HANDLE *f, char *t, const int start)
 		return -1;
 
 	n = hio_read8(f);			/* instruments */
-	if (n > 100)
+	if (n <= 0 || n > 100)
 		return -1;
 
 	hio_read8(f);			/* sequences */
@@ -82,9 +82,11 @@ static int coco_test(HIO_HANDLE *f, char *t, const int start)
 		if (ofs < 64 || ofs > 0x00100000)
 			return -1;
 
-		if (vol > 0xff)
+		if (vol < 0 || vol > 0xff)
 			return -1;
 
+		if (len < 0 || lps < 0 || lsz < 0)
+			return -1;
 		if (len > 0x00100000 || lps > 0x00100000 || lsz > 0x00100000)
 			return -1;
 
@@ -203,6 +205,10 @@ static int coco_load(struct module_data *m, HIO_HANDLE *f, const int start)
 	seq_ptr = hio_read32l(f);
 	pat_ptr = hio_read32l(f);
 
+	if (hio_error(f)) {
+		return -1;
+	}
+
 	MODULE_INFO();
 
 	if (libxmp_init_instrument(m) < 0)
@@ -235,6 +241,10 @@ static int coco_load(struct module_data *m, HIO_HANDLE *f, const int start)
 		if (mod->xxs[i].len > 0)
 			mod->xxi[i].nsm = 1;
 
+		if (hio_error(f)) {
+			return -1;
+		}
+
 		D_(D_INFO "[%2X] %-10.10s  %05x %05x %05x %c V%02x",
 				i, mod->xxi[i].name,
 				mod->xxs[i].len, mod->xxs[i].lps, mod->xxs[i].lpe,
@@ -254,7 +264,6 @@ static int coco_load(struct module_data *m, HIO_HANDLE *f, const int start)
 	for (i++; i % 4; i++)	/* for alignment */
 		hio_read8(f);
 
-
 	/* Patterns */
 
 	if (libxmp_init_pattern(mod) < 0)
```

**Comment by AliceLR** [comment]
On top of the checks recommended by @sezero it might also be a good idea to add an `hio_error` check in the pattern loading loop, since that's usually where the most time can be wasted reading EOFs.

Aside from those, overall the changes here look good (the `fix_effect` comments are especially appreciated). Re: unimplemented effects, these might not be entirely what Coconizer needs but are still worth looking into:

| Coconizer | Possible libxmp mapping |
|-|-|
| `08` auto fine volslide up | `FX_PER_VSLD_UP` (unfortunately looks like regular slide, not fine) |
| `09` auto fine volslide down | `FX_PER_VSLD_DN` (same) |
| `0A` auto portamento up | `FX_PER_PORTA_UP` |
| `0B` auto portamento down | `FX_PER_PORTA_DN` |
| `10` fine portamento up | `FX_F_PORTA_UP` |
| `11` fine portamento down | `FX_F_PORT_DN` |

I haven't made enough sense of libxmp's effects mappings yet to have any recommendations for how to add completely new effects. (IMO having a common effects mapping like libxmp does leads to subtle bugs and it'd be better to have individual format effect mapping functions call into common effect functions that can be easily duplicated/etc for compatibility. That would be a lot of work...)

File: src/loaders/coco_load.c [hunk 1]
```diff
@@ -275,6 +284,10 @@ static int coco_load(struct module_data *m, HIO_HANDLE *f, const int start)
 			if (event->note)
 				event->note += 12;
 
+			if (hio_error(f)) {
```
Comment (sezero) [inline src/loaders/coco_load.c:287]: Nitpick: _maybe_ move this just before the `event->note` handling above?

In any case, this looks better.  One place I'm not sure is whether to check
the read `mod->chn` value being positive around line 190: Many other loaders
do not seem to check, and _maybe_ a bad value would be handled elsewhere
(didn't look close enough, though.)

Comment (AliceLR) [reply src/loaders/coco_load.c:287]: The value read for `mod->chn` is an unsigned byte and is bounded to the valid channels range with `& 0x3f`, which is really tacky but technically works (`XMP_MAX_CHANNELS` is 64). The file format document suggests it should only ever be 4 or 8 but I've seen format documents blatantly lie before...

Comment (sezero) [reply src/loaders/coco_load.c:287]: OK.  Leaving fine-tuning and final review to you.

**Comment by sezero** [comment]
Any progress on this?

**Comment by AliceLR** [comment]
> Any progress on this?

Finishing up reviewing now and then I'll go ahead and merge it. I'll add a format test as a separate pull request.

**Review by AliceLR — APPROVED** [review]
This looks good to me. Every module in "e503" from APDL loads as far as I can tell. "Poseidon" seems to have some portamento bugs but those are almost certainly not caused by this branch.

@ccawley2011 The test modules in the "e501" archive aren't loaded and (when depacked via the ArcFS depacker) appear to have an extra header at the start which includes the string `CoconizerSong`. The Coconizer data libxmp currently expects starts a bit further in the file (offset 2792 in the case of "Beethoven") and doesn't have the trackfile flag set. Despite that, there does appear to be sample data in later these files. Any idea what's going on with these?

**Comment by ccawley2011** [comment]
> @ccawley2011 The test modules in the "e501" archive aren't loaded and (when depacked via the ArcFS depacker) appear to have an extra header at the start which includes the string `CoconizerSong`. The Coconizer data libxmp currently expects starts a bit further in the file (offset 2792 in the case of "Beethoven") and doesn't have the trackfile flag set. Despite that, there does appear to be sample data later in these files. Any idea what's going on with these?

The tracks in "e501" are embedded in relocatable modules containing the player code, which can be converted to standard Coconizer format using the `!Toolbox` application in "e502". I've attached an archive containing converted versions of the tracks in "e501", which includes files that use "Start Auto Fine Volume Up/Down" and "Slide Pitch Up/Down (0 limit)".

[Module.tar.gz](https://github.com/libxmp/libxmp/files/6329347/Module.tar.gz)

**Pull request merged**